applications/ChimeraApplication/custom_utilities/hole_cutting_utility.cpp

```diff
@@ -0,0 +1,438 @@
+        else if (TDim == 3)
+            faces = i_element->GetGeometry().GenerateFaces();
+
+        for (IndexType i_face = 0; i_face < faces.size(); i_face++) {
+            // Create vector that stores all node is of current i_face
+            vector<IndexType> ids(faces[i_face].size());
+            vector<IndexType> unsorted_ids(faces[i_face].size());
+
+            // Store node ids
+            for (IndexType i = 0; i < faces[i_face].size(); i++) {
+                ids[i] = faces[i_face][i].Id();
+                unsorted_ids[i] = faces[i_face][i].Id();
+            }
+
+            //*** THE ARRAY OF IDS MUST BE ORDERED!!! ***
+            std::sort(ids.begin(), ids.end());
+#pragma omp critical
+            {
+                if (n_faces_map[ids] == 1)
+                    ordered_skin_face_nodes_map[ids] = unsorted_ids;
+            }
+        }
+    }
+    // First assign to skin model part all nodes from original model_part,
+    // unnecessary nodes will be removed later
+    IndexType id_condition = 1;
+
+    // Add skin faces as triangles to skin-model-part (loop over all node sets)
+    std::vector<IndexType> vector_of_node_ids;
+    for (typename hashmap::const_iterator it = n_faces_map.begin();
+         it != n_faces_map.end(); ++it) {
+        // If given node set represents face that is not overlapping with a face
+        // of
+        // another element, add it as skin element
+        if (it->second == 1) {
+            // If skin edge is a triangle store triangle in with its original
+            // orientation in new skin model part
+            if (it->first.size() == 2) {
+                // Getting original order is important to properly reproduce
+                // skin edge
+                // including its normal orientation
+                vector<IndexType> original_nodes_order =
+                    ordered_skin_face_nodes_map[it->first];
+
+                Node<3>::Pointer pnode1 =
+                    rVolumeModelPart.Nodes()(original_nodes_order[0]);
+                Node<3>::Pointer pnode2 =
+                    rVolumeModelPart.Nodes()(original_nodes_order[1]);
+
+                // Storing the node ids list
+                vector_of_node_ids.push_back(original_nodes_order[0]);
+                vector_of_node_ids.push_back(original_nodes_order[1]);
+
+                Properties::Pointer properties =
+                    rExtractedBoundaryModelPart.rProperties()(0);
+                Condition const& rReferenceLineCondition =
+                    KratosComponents<Condition>::Get("LineCondition2D2N"); // Condition2D
+
+                // Skin edges are added as conditions
+                Line2D2<Node<3>> line1(pnode1, pnode2);
+                Condition::Pointer p_condition1 =
+                    rReferenceLineCondition.Create(id_condition++, line1, properties);
+                rExtractedBoundaryModelPart.Conditions().push_back(p_condition1);
```

Comment:
ouchouchouch
You should use the ModelPart interface otherwise you break the structure of it!
Best to collect the conditions to be added in a vector and add them add once
This is often done in processes in the core (don't know by hard which ones)

Reply:
This I used to do previously. This proved very very very expensive. This is because there is and find and check for every condition that we want to add. (even when added from a vector/container). For smaller cases this is not a problem, but as soon as the case is little bigger, you see the time taken is quickly very big.

Reply:
hm ok :confused: 

@RiccardoRossi @pooyan-dadvand this is probably sth that should get fixed/improved
this affects many other places too!

@adityaghantasala could you please open an issue abt this? I think this is quite important

BTW this also proves the point I many in other places abt using Ids. Exactly same issue